ISPN-8097 ScatteredCrashInSequenceTest.testSplit failures

rvansa opened the pull request:
https://issues.jboss.org/browse/ISPN-8097

rvansa commented:
(not sure if the test fix label applies, since it's not a test problem)

pruivo commented:
basically, the only change it setting the topology id in the get. It looks good to me :) 
IMO, the test-fix label applies if you are fixing a test that is randomly (or not) failing :)

pruivo commented:
there are still failures in CI `ScatteredCrashInSequenceTest`

danberindei commented on core/src/main/java/org/infinispan/interceptors/impl/PrefetchInterceptor.java:
```diff
@@ -207,68 +214,97 @@ private Object prefetchKeysIfNeededAndInvokeNext(InvocationContext ctx, Visitabl
+      // Moreover, OTE thrown this way has requested topology id set; all queries about owner etc. must match
+      // command's topology id. If it did not the cache topology could change between SDI and STI
+      // (when the exception pops up) and we would set next topology id too high, while the OTE/AOLE was thrown
+      // due to older topology being present in SDI.
```
I'd rather point the reader to the `ScatteredDistributionInterceptor` (?) method that checks the topology id and throws the OTE instead. This looks like too much detail and too far from the implementation to stay true for long.

danberindei commented on core/src/main/java/org/infinispan/interceptors/impl/PrefetchInterceptor.java:
```diff
@@ -207,68 +214,97 @@ private Object prefetchKeysIfNeededAndInvokeNext(InvocationContext ctx, Visitabl
    }
 
-   private InvocationStage retrieveSingleRemoteValue(InvocationContext ctx, Object key) {
+   private InvocationStage retrieveSingleRemoteValue(InvocationContext ctx, Object key, int topologyId) {
```
Is the method name right? At the very least, it's too similar to `retrieveRemoteValue`, which is quite different.

rvansa replied:
Right, it sounds a bit confusing. `retrieveSingleRemoteValue` runs first the local lookup (for the case that we already have the key) and then invokes `retrieveRemoteValue`. How would you like to call this method which does first local and then remote lookup? Just `retrieveSingleValue`?

Writing this comment I've realized that it may not work properly, though; Using the RemoteMetadata is correct only during the value transfer, but during key transfer the metadata could point to a node with outdated copy. I'll change that.

danberindei replied:
I'd probably call it `invokeGetCacheEntryCommand`.

rvansa replied:
That (naming) wouldn't be right: it invokes the command *and* then does the remote lookup. The returned InvocationStage is completed when both those complete.

danberindei commented on core/src/main/java/org/infinispan/interceptors/impl/PrefetchInterceptor.java:
```diff
@@ -207,68 +214,97 @@ private Object prefetchKeysIfNeededAndInvokeNext(InvocationContext ctx, Visitabl
+         log.tracef("Prefetching entries for keys %s using broadcast", keys);
+      }
       ClusteredGetAllCommand command = commandsFactory.buildClusteredGetAllCommand(keys, FlagBitSets.SKIP_OWNERSHIP_CHECK, null);
+      command.setTopologyId(topologyId);
```
No explanation needed here?

rvansa replied:
TBH I haven't seen any problems with that but I thought that adding the id won't harm. Should i add `// not really tested`? :)

danberindei replied:
It's just odd to have such a long explanation in one method and no explanation in the another one when they do the same thing :)

TBH it's probably better to not have a comment in `PrefetchInterceptor`, and have `ScatteredDistributionInterceptor` throw an exception if the topology id is not set. In https://github.com/infinispan/infinispan/pull/5315 I also made `RpcManager.invokeCommand*()` throw an exception if the topology id is not set instead of using the current topology id, like `RpcManager.invokeRemotely()` does.

rvansa commented:
I've been running the tests 100 times w/ trace logs and 12 times w/o logs without failure :-/

rvansa commented:
Updated, setting the topology id all the time, including for local commands (those could do the retry now, too - I think that it's little price for reduced complexity).

danberindei commented:
@rvansa I don't think it's conceptually ok to throw an OTE for `CACHE_MODE_LOCAL` commands, because you're not doing an RPC, but you're still waiting for another RPC to happen so that the new topology is installed.

There's also a problem with `throw new OutdatedTopologyException(command.getTopologyId())` in `ScatteredDistributionInterceptor.checkTopology()`. There's an undocumented assumption in `(Base)StateTransferInterceptor` that the `requestedTopologyId` of any caught OTE is always greater than the command's initial topology id, and some of the `handleXReturn` methods ignore it, using `command.getTopologyId() + 1` all the time.

OTOH the `(Base)StateTransferInterceptor.handleXReturn` methods that do not ignore the `requestedTopologyId` won't block at all, because the command's initial topology id is guaranteed to be less than the current topology id.

And finally, what does it mean for an exception to "pop up"?

rvansa commented:
I agree that you must not wait for RPC with CACHE_MODE_LOCAL command, but I thought that the OTE is thrown only if the topology is changed between STI and xDI. Then we should retry immediatelly, because the next topology should be already installed. You can't have OTEs originating remotely.

Though, there really seems to be something fishy about BaseSTI.nextTopology and getNextTopologyId, I'll check it out.

Exception popping up means being propagated up the stack. I'll change the wording.

danberindei commented:
@rvansa I was thinking about when the primary owner is `null`, but I guess that would throw an AOLE instead of an OTE. Either way, a `CACHE_MODE_LOCAL` command should not wait for a new topology with a non-null primary owner.

rvansa commented:
@danberindei OK, I'll just remove the topologyId check for local commands in distribution interceptors, does that sound good to you?

tristantarrant commented:
Looks like this is still failing http://ci.infinispan.org/job/Infinispan/job/PR-5317/7/testReport/junit/org.infinispan.partitionhandling/ScatteredCrashInSequenceTest/testSplit5_SCATTERED_SYNC_/

rvansa commented:
@tristantarrant Yes, I've noticed that CI is still failing. So this will be a fix for the problems that I can reproduce locally.

danberindei commented:
@rvansa Skipping the topology check for local commands sounds good. I'll keep the JIRA open when I integrate this :)

pruivo commented:
@rvansa I have the `testSplit2` failing when I run the test class with single thread.

pruivo commented:
trace logs of `testSplit4`
https://drive.google.com/file/d/0ByatUPYii_o5UVpEWHFfZnFJekk/view?usp=sharing

I've noticed that NodeC receives an InvalidateVersionCommand right after applied the state transfer... I'm not sure if it was supposed to do that or not...

rvansa commented:
@pruivo Looks like you've hit even different issue than the one in CI. I'll fix that as well.

rvansa commented:
Okay, I've fixed both the InvalidateVersionsCommand coming from the past topology, and the other issue that had troubled CI: the test was using `StateSequencer` to block rebalance confirmations, but if none of them came it got stuck as well. Replaced with one-way synchronization.

pruivo commented:
@rvansa checkstyle error in CI

rvansa commented on core/src/main/java/org/infinispan/interceptors/distribution/ScatteredDistributionInterceptor.java:
```diff
@@ -714,7 +712,7 @@ private Object handlePutMapOnOrigin(InvocationContext ctx, PutMapCommand command
             commitSingleEntryIfNewer(ctxEntry, ctx, command);
          } else if (info.primary() == null) {
-            throw AllOwnersLostException.INSTANCE;
+            throw new AllOwnersLostException(cacheTopology.getTopologyId());
```
There's no owner in CH - the key is lost.

danberindei replied:
The way I see it, a key is lost iif `cache.get(key)` returns `null` or throws an `AvailabilityException`. Scattered cache may have a single owner in the CH, but it behaves as if it had 1 primary owner and 1 backup owner.

rvansa replied:
Still, by definition of "all owners" this is what happens. I prefer to throw OTE on events where we detect that current topology id does not match to command topology id, rather than any time we don't know what else to do and hope we could know it later.

danberindei replied:
I guess I'm fixating too much on the "lost" part :)

But IMO we don't need a separate exception every time we need to tell `StateTransferInterceptor` to wait for a new topology, because that just adds duplicate code in `StateTransferInterceptor`. Except here the topology id in `AllOwnersLostException` is the "known bad" topology id and in `OutdatedTopologyException` it's the "expected" topology id, so it's going to be harder to figure out if `STI` is trying to do the same thing for both in the future.

rvansa commented on core/src/main/java/org/infinispan/interceptors/impl/BaseStateTransferInterceptor.java:
```diff
@@ -237,13 +232,24 @@ private Object handleReadCommandReturn(InvocationContext rCtx, VisitableCommand
+            AllOwnersLostException aole = (AllOwnersLostException) ce;
+            // It is possible that a command arrives with higher topology id than the one of current node; such
+            // command can cause execution of another command with CACHE_MODE_LOCAL flag (not checking topology id
+            // vs. command id) and this could find no owner in the current cache topology.
```
It was your requirement that a read command never blocks, waiting for another topology.

danberindei replied:
That was for distributed mode. I thought scattered mode always had to block until the primary (as determined by the originator) has the value, and the primary can't apply the value if it doesn't have the right topology.

rvansa commented on core/src/main/java/org/infinispan/interceptors/impl/PrefetchInterceptor.java:
```diff
@@ -207,68 +214,96 @@ private Object prefetchKeysIfNeededAndInvokeNext(InvocationContext ctx, Visitabl
+         return null;
+      } else if ((metadata instanceof RemoteMetadata) && svm.getSegmentState(segment) == ScatteredVersionManager.SegmentState.VALUE_TRANSFER) {
+         // The RemoteMetadata is valid only during value transfer - in blocked state there shouldn't be any such
+         // entry and during key transfer we could see metadata pointing to a node with outdated information.
```
It's not outdated RemoteMetadata, but RM pointing to an outdated copy. During key transfer all nodes send their versions and RM is added/overwritten if another node has higher version. Until we get all copies (and we're not sure about it until value transfer) there might be a RM that comes from a node with outdated information.

danberindei replied:
Maybe you could add some javadoc to `RemoteMetadata` to explain what `getAddress()` means during each phase?

rvansa replied:
added

rvansa commented on core/src/main/java/org/infinispan/commands/write/InvalidateVersionsCommand.java:
```diff
@@ -38,39 +45,63 @@ public InvalidateVersionsCommand(ByteString cacheName) {
+               }
+            });
+         }
+         return orderedUpdatesManager.invalidate(keys).handle((nil, t) -> {
```
what's the difference between DC and storing in persistence?

danberindei replied:
I wasn't thinking about persistence, but about not blocking while holding the shared topology lock in general.

`StateTransferLockImpl.ownershipLock` is non-fair, so keeping the read lock for a long time may delay write lock acquisition (IOW topology updates) for a long time. Of course, it may be simpler to make `ownershipLock` fair, but then we'd have to profile it and make sure it's not slower than then non-fair version when no locks are acquired.

rvansa commented:
@pruivo @danberindei Forgot to comment, the checkstyle problems have been fixed and now there's Success in CI (First time I see that!)

pruivo commented:
@rvansa success? I see a new failure:
http://ci.infinispan.org/job/Infinispan/job/PR-5317/24/testReport/org.infinispan.scattered.statetransfer/PushTransferTest/testNodeJoin/

rvansa commented:
@pruivo @danberindei Ok, this time it asked for a bigger overhaul. I've found out that rebalance confirmations can be sent before the push-transfer (incl. invalidation) finishes, so I've adjusted `StateTransferProvider/Consumer#onTopology` to return CF and confirm the rebalance only after both complete the update, instead of calling `StateTransferManager.notifyRebalanceEnd` which I've removed.

danberindei commented on core/src/main/java/org/infinispan/container/entries/RemoteMetadata.java:
```diff
@@ -14,6 +14,11 @@
+ * This is a metadata type used by scattered cache during state transfer. The address points to node which has last
+ * known version of given entry: During key transfer RemoteMetadata is created and overwritten if another response
+ * with higher version comes. During value transfer the address is already final and we request the value + metadata
+ * only from this node.
```
Nice, this is exactly what I was looking for!

danberindei commented on core/src/main/java/org/infinispan/interceptors/impl/PrefetchInterceptor.java:
```diff
@@ -207,68 +214,96 @@ private Object prefetchKeysIfNeededAndInvokeNext(InvocationContext ctx, Visitabl
+         Metadata metadata = icv.getMetadata();
+         if (metadata instanceof RemoteMetadata) {
+            // not sure if this can happen, but let's be on the safe side
+            throw OutdatedTopologyException.INSTANCE;
```
I think it can happen if the remote get is delayed a lot and the key ownership changes from A to B and then to A again.

rvansa replied:
Right, topology check should make this impossible.

danberindei commented:
Integrated, thanks Radim!
